stress: failed test in cockroach/storage/storage.test: TestStoreMetrics

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/b1125fa08614aa770703b67a18dfb9856e2cd4ad

Stress build found a failed test:

```
=== RUN   TestStoreMetrics
W160313 07:34:01.228165 gossip/gossip.go:907  not connected to cluster; use --join to specify a connected node
W160313 07:34:01.229018 gossip/gossip.go:907  not connected to cluster; use --join to specify a connected node
I160313 07:34:01.229581 storage/replica_command.go:1293  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1.000s
W160313 07:34:01.230010 gossip/gossip.go:907  not connected to cluster; use --join to specify a connected node
I160313 07:34:01.232304 gossip/gossip.go:933  starting client to 127.0.0.1:44164
W160313 07:34:01.232501 gossip/gossip.go:907  not connected to cluster; use --join to specify a connected node
I160313 07:34:01.236169 gossip/gossip.go:933  starting client to 127.0.0.1:44164
I160313 07:34:01.237984 storage/replica_command.go:1611  initiating a split of range=1 [/Min-/Max) at key "m"
I160313 07:34:01.240004 storage/replica_command.go:1293  range 2: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1.000s
W160313 07:34:01.240507 storage/stores.go:207  range not contained in one range: [/Meta2/Max,"m\x00"), but have [/Min,"m")
I160313 07:34:01.242108 storage/replica.go:1138  raft: proposing ADD_REPLICA {2 2 2} for range 2
W160313 07:34:01.242192 storage/replica.go:1856  [node=1,store=1,range=1] failed to push during intent resolution: conflicting intents on [/Meta2/Max]: resolved? false
I160313 07:34:01.246483 storage/replica.go:1138  raft: proposing ADD_REPLICA {3 3 3} for range 2
E160313 07:34:01.254909 storage/raft_transport.go:229  rpc error: code = 13 desc = "transport: write tcp 127.0.0.1:40852->127.0.0.1:44164: use of closed network connection"
--- FAIL: TestStoreMetrics (0.03s)
	client_metrics_test.go:47: storage/client_metrics_test.go:84: sysbytes for store: actual 477 != expected 664
	client_metrics_test.go:47: storage/client_metrics_test.go:85: syscount for store: actual 6 != expected 7
	client_metrics_test.go:91: storage/client_metrics_test.go:121: verifyStats failed, aborting test.
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
1 runs so far, 0 failures, over 20s
2 runs so far, 0 failures, over 25s
8 runs so far, 0 failures, over 30s
8 runs so far, 0 failures, over 35s
8 runs so far, 0 failures, over 40s
9 runs so far, 0 failures, over 45s
11 runs so far, 0 failures, over 50s
16 runs so far, 0 failures, over 55s
16 runs so far, 0 failures, over 1m0s
17 runs so far, 0 failures, over 1m5s
17 runs so far, 0 failures, over 1m10s
19 runs so far, 0 failures, over 1m15s
24 runs so far, 0 failures, over 1m20s
24 runs so far, 0 failures, over 1m25s
25 runs so far, 0 failures, over 1m30s
25 runs so far, 0 failures, over 1m35s
27 runs so far, 0 failures, over 1m40s
32 runs so far, 0 failures, over 1m45s
32 runs so far, 0 failures, over 1m50s
33 runs so far, 0 failures, over 1m55s
33 runs so far, 0 failures, over 2m0s
36 runs so far, 0 failures, over 2m5s
40 runs so far, 0 failures, over 2m10s
40 runs so far, 0 failures, over 2m15s
41 runs so far, 0 failures, over 2m20s
41 runs so far, 0 failures, over 2m25s
44 runs so far, 0 failures, over 2m30s
47 runs so far, 0 failures, over 2m35s
47 runs so far, 0 failures, over 2m40s
48 runs so far, 0 failures, over 2m45s
48 runs so far, 0 failures, over 2m50s
54 runs so far, 0 failures, over 2m55s
54 runs so far, 0 failures, over 3m0s
54 runs so far, 0 failures, over 3m5s
55 runs so far, 0 failures, over 3m10s
55 runs so far, 0 failures, over 3m15s
61 runs so far, 0 failures, over 3m20s
61 runs so far, 0 failures, over 3m25s
61 runs so far, 0 failures, over 3m30s
63 runs so far, 0 failures, over 3m35s
63 runs so far, 0 failures, over 3m40s
69 runs so far, 0 failures, over 3m45s
69 runs so far, 0 failures, over 3m50s
70 runs so far, 0 failures, over 3m55s
71 runs so far, 0 failures, over 4m0s
72 runs so far, 0 failures, over 4m5s
77 runs so far, 0 failures, over 4m10s
77 runs so far, 0 failures, over 4m15s
78 runs so far, 0 failures, over 4m20s
80 runs so far, 0 failures, over 4m25s
80 runs so far, 0 failures, over 4m30s
81 runs completed, 1 failures, over 4m31s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/storage/client_metrics_test.go
+++ b/storage/client_metrics_test.go
@@ -54,7 +54,17 @@ func checkCounter(t *testing.T, s *storage.Store, key string, e int64) {
 	}
 }
 
-func verifyStats(t *testing.T, s *storage.Store) {
+func verifyStats(t *testing.T, mtc *multiTestContext, storeIdx int) {
+	// Get the current store at storeIdx.
+	s := mtc.stores[storeIdx]
+	// Stop the store at the given index, while keeping the the reference to the
+	// store object. Method ComputeMVCCStatsTest() still works on a stopped
+	// store (it needs only the engine, which is still open), and the most
+	// recent stats are still available on the stopped store object; however, no
+	// further information can be committed to the store while it is stopped,
+	// preventing any races during verification.
+	mtc.stopStore(storeIdx)
+
 	// Compute real total MVCC statistics from store.
 	realStats, err := s.ComputeMVCCStatsTest()
 	if err != nil {
@@ -91,6 +101,9 @@ func verifyStats(t *testing.T, s *storage.Store) {
 		t.Log(util.ErrorfSkipFrames(1, "verifyStats failed, aborting test."))
 		t.FailNow()
 	}
+
+	// Restart the store at the provided index.
+	mtc.restartStore(storeIdx)
 }
 
 func verifyRocksDBStats(t *testing.T, s *storage.Store) {
@@ -127,37 +140,34 @@ func TestStoreMetrics(t *testing.T) {
 	mtc := startMultiTestContext(t, 3)
 	defer mtc.Stop()
 
-	store0 := mtc.stores[0]
-	store1 := mtc.stores[1]
-
 	// Flush RocksDB memtables, so that RocksDB begins using block-based tables.
 	// This is useful, because most of the stats we track don't apply to
 	// memtables.
-	if err := store0.Engine().Flush(); err != nil {
+	if err := mtc.stores[0].Engine().Flush(); err != nil {
 		t.Fatal(err)
 	}
-	if err := store1.Engine().Flush(); err != nil {
+	if err := mtc.stores[1].Engine().Flush(); err != nil {
 		t.Fatal(err)
 	}
 
 	// Perform a split, which has special metrics handling.
 	splitArgs := adminSplitArgs(roachpb.KeyMin, roachpb.Key("m"))
-	if _, err := client.SendWrapped(rg1(store0), nil, &splitArgs); err != nil {
+	if _, err := client.SendWrapped(rg1(mtc.stores[0]), nil, &splitArgs); err != nil {
 		t.Fatal(err)
 	}
 
 	// Verify range count is as expected
-	checkCounter(t, store0, "ranges", 2)
+	checkCounter(t, mtc.stores[0], "ranges", 2)
 
 	// Verify all stats on store0 after split.
-	verifyStats(t, store0)
+	verifyStats(t, mtc, 0)
 
 	// Replicate the "right" range to the other stores.
-	replica := store0.LookupReplica(roachpb.RKey("z"), nil)
+	replica := mtc.stores[0].LookupReplica(roachpb.RKey("z"), nil)
 	mtc.replicateRange(replica.RangeID, 1, 2)
 
 	// Verify stats on store1 after replication.
-	verifyStats(t, store1)
+	verifyStats(t, mtc, 1)
 
 	// Add some data to the "right" range.
 	dataKey := []byte("z")
@@ -167,8 +177,8 @@ func TestStoreMetrics(t *testing.T) {
 	mtc.waitForValues(roachpb.Key("z"), []int64{5, 5, 5})
 
 	// Verify all stats on store 0 and 1 after addition.
-	verifyStats(t, store0)
-	verifyStats(t, store1)
+	verifyStats(t, mtc, 0)
+	verifyStats(t, mtc, 1)
 
 	// Create a transaction statement that fails, but will add an entry to the
 	// sequence cache. Regression test for #4969.
@@ -181,7 +191,8 @@ func TestStoreMetrics(t *testing.T) {
 	}
 
 	// Verify stats after sequence cache addition.
-	verifyStats(t, store0)
+	verifyStats(t, mtc, 0)
+	checkCounter(t, mtc.stores[0], "ranges", 2)
 
 	// Unreplicate range from the first store.
 	mtc.unreplicateRange(replica.RangeID, 0)
@@ -191,13 +202,13 @@ func TestStoreMetrics(t *testing.T) {
 	mtc.waitForValues(roachpb.Key("z"), []int64{0, 5, 5})
 
 	// Verify range count is as expected.
-	checkCounter(t, store0, "ranges", 1)
-	checkCounter(t, store1, "ranges", 1)
+	checkCounter(t, mtc.stores[0], "ranges", 1)
+	checkCounter(t, mtc.stores[1], "ranges", 1)
 
 	// Verify all stats on store0 and store1 after range is removed.
-	verifyStats(t, store0)
-	verifyStats(t, store1)
+	verifyStats(t, mtc, 0)
+	verifyStats(t, mtc, 1)
 
-	verifyRocksDBStats(t, store0)
-	verifyRocksDBStats(t, store1)
+	verifyRocksDBStats(t, mtc.stores[0])
+	verifyRocksDBStats(t, mtc.stores[1])
 }

